File: python/day05.py
from collections.abc import Iterator
from typing import Any


class Page:
    def __init__(self, n: str, rules: set[tuple[int, int]]) -> None:
        self.n = int(n)
        self.rules = rules

    def __eq__(self, other: Any) -> bool:
        assert isinstance(other, Page)
        return self.n == other.n

    def __lt__(self, other: "Page") -> bool:
        return (self.n, other.n) in self.rules


def extract_rules(text: str) -> set[tuple[int, int]]:
    rules = set()
    for line in text.splitlines():
        n1, n2 = line.split("|", maxsplit=1)
        rule = int(n1), int(n2)
        rules.add(rule)
    return rules


def extract_pages(text: str, rules: set[tuple[int, int]]) -> Iterator[list[Page]]:
    for line in text.splitlines():
        yield [Page(n, rules) for n in line.split(",")]


def run(text: str) -> tuple[Any, Any]:
    rule_text, page_text = text.split("\n\n", maxsplit=1)
    rules = extract_rules(rule_text)

    sum1 = 0
    sum2 = 0
    for pages in extract_pages(page_text, rules):
        middle = len(pages) // 2
        ordered = sorted(pages)
        if ordered == pages:
            sum1 += ordered[middle].n
        else:
            sum2 += ordered[middle].n

    return sum1, sum2

File: zig/day05.zig
const std = @import("std");
const Allocator = std.mem.Allocator;
const ArrayList = std.ArrayList;
const AutoHashMap = std.AutoHashMap;

fn extract_rules(input: []const u8, rules: *AutoHashMap([2]u8, void)) !void {
    var rules_iter = std.mem.tokenizeScalar(u8, input, '\n');
    while (rules_iter.next()) |rule_text| {
        const a = try std.fmt.parseInt(u8, rule_text[0..2], 10);
        const b = try std.fmt.parseInt(u8, rule_text[3..5], 10);
        try rules.put(.{ a, b }, {});
    }
}

fn extract_pages(input: []const u8, pages: *ArrayList(u8)) !void {
    var page_iter = std.mem.tokenizeScalar(u8, input, ',');
    while (page_iter.next()) |page_text| {
        const page = try std.fmt.parseInt(u8, page_text, 10);
        try pages.append(page);
    }
}

fn sorter(rules: *AutoHashMap([2]u8, void), lhs: u8, rhs: u8) bool {
    if (rules.get(.{ lhs, rhs })) |_| {
        return true;
    }
    return false;
}

pub fn main(allocator: Allocator, input: []const u8) !void {
    var input_iter = std.mem.splitSequence(u8, input, "\n\n");
    const rules_input = input_iter.next().?;
    const pages_input = input_iter.next().?;

    var rules = AutoHashMap([2]u8, void).init(allocator);
    defer rules.deinit();
    try extract_rules(rules_input, &rules);

    var pages = ArrayList(u8).init(allocator);
    defer pages.deinit();

    var sorted = ArrayList(u8).init(allocator);
    defer sorted.deinit();

    var part1: u32 = 0;
    var part2: u32 = 0;

    var pages_iter = std.mem.tokenizeScalar(u8, pages_input, '\n');
    while (pages_iter.next()) |pages_text| {
        pages.clearRetainingCapacity();
        try extract_pages(pages_text, &pages);

        sorted.clearRetainingCapacity();
        try sorted.appendSlice(pages.items);

        std.mem.sort(u8, sorted.items, &rules, sorter);
        const middle = sorted.items[@divTrunc(sorted.items.len, 2)];

        if (std.mem.eql(u8, pages.items, sorted.items)) {
            part1 += middle;
        } else {
            part2 += middle;
        }
    }

    std.debug.print("{}\n", .{part1});
    std.debug.print("{}\n", .{part2});
}
